---
title: "Case-Control genome-wide association tutorial"
author: "Jason Moggridge"
date: "`r Sys.Date()`"
output:
  pdf_document:
    highlight: kate
    toc: true
    toc_depth: 3
urlcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message =F, warning = F, cache = T)
```

---

# Overview:

This tutorial deals with performing a statistical analysis of SNP
data for a case-control genome-wide association study using the popular and
versatile [`plinck` toolset](https://zzz.bwh.harvard.edu/plink/index.shtml).
We will perform quality control, exploratory analysis of data with MDS, then test for genome-wide association, and plot the results using R.
The sample dataset is from a recent GWAS in Labrador retrievers with a relatively small number of individuals.

## Process outline:

 - 1 Obtain data and load modules
 - 2 QC + filtering data
 - 3 Pop stratification, MDS / clustering
 - 4 Association tests and statistics

 
## Dataset: 

This tutorial uses a publicly-available dataset with a manageable size, 173,662 SNPs and 237 dogs, from a case-control GWAS for ACL rupture, which is fairly common in Labradors. 
  
**Related publication**: 
Genome-wide association analysis in dogs implicates 99 loci as risk variants for anterior cruciate ligament rupture.  
Baker LA, Kirkpatrick B, Rosa GJM, Gianola D, Valente B, et al. (2017) *PLOS ONE* 12(4): e0173810. 
https://doi.org/10.1371/journal.pone.0173810
  
  
**Data source**: Data dryad link  https://datadryad.org/stash/dataset/doi:10.5061/dryad.8kk06 (We'll download the data with `curl` in the tutorial)

**Files**:  
 - `cr237_dryad.bed`: has genotype data   
 - `cr237_dryad.bim`: has info about each variant (human-readable)    
 - `cr237_dryad.fam`: has phenotype data (human-readable)    
  



## Plink usage primer

    plink --bfile {my_data} --dog --options --commands --out {new_name} 

 Throughout this tutorial, we use the `--bfile` flag because we have bed/bim/fam files; the `--make-bed` flag to get the same type of files back; and we always have to use the `--dog` flag because dogs have 38 chromosomes and plink expects human data otherwise (for every command!).  
For various filtering steps, plink has `--keep` and `--remove` commands for filtering individuals, while the `--extract` and `--exclude` functions do the same for variants. Each of these requires a text file with a list of identifiers, eg. <`--keep keep_list.txt`> .


---

# Step-by-step tutorial

`plink` commands are run in the unix shell. Later on we will do some plotting in R.


## 1 - Data and software

Let's get the dataset from datadryad. Do this on a login node on `graham`. Download & unzip. The files are relatively small. (If the `curl` throws an error, try exiting graham and logging in again)

```{r download, engine = 'bash', eval = F, include = T}
curl -L http://datadryad.org/api/v2/datasets/doi%253A10.5061%252Fdryad.8kk06/download \
  --output labrador_download
unzip labrador_download
```

We got the cr237_dryad.bed, .bim., and .fam files:

  - .bed is a binary file with genotypes
  - .bim is a binary file with info about variant loci
  - .fam has the phenotype info and genders and is text

The .fam file has no header, so it's useful to know the columns that have the identifiers, sex and phenotype data for the 237 dogs:

    1 Family ID ('FID')
    2 Within-family ID ('IID'; cannot be '0')
    3 Within-family ID of father ('0' if father isn't in dataset)
    4 Within-family ID of mother ('0' if mother isn't in dataset)
    5 Sex code ('1' = male, '2' = female, '0' = unknown)
    7 Phenotype value ('1' = control, '2' = case, '-9'/'0'/non-numeric = missing data if case/control)


```{r famfile, engine = 'bash', eval = F, include = T}
head cr237_dryad.fam
```

There are no relatives in this dataset, so the FIDs are all the same as the IIDs, and columns 3 and 4 are all zeros. Our binary trait of interest (ACL rupture) is in column 6.   
See the plink documentation for more info about file formats: https://www.cog-genomics.org/plink/1.9/formats (other input formats are possible, if you're working with files that don't match this pattern).


Start an srun session, then load these modules:

```{r modules, engine = 'bash', eval = F, include = T}
module load nixpkgs/16.09 gcc/7.3.0 r/4.0.2 plink/1.9b_4.1-x86_64 
```

----

## 2 - QC / Filtering


### a. Data missingness

First we'll check the amount of missing data per individual with `--missing`.

```{r missing, engine = 'bash', eval = F, include = T}
plink --bfile cr237_dryad --missing --dog
```

The files .imiss .lmiss show the proportion of missing data per loci and indiv. Download these to your local machine and check the distribution if you want to adjust the thresholds. There is code for plotting these in the R section at the end of the tutorial.

Here, we'll filter any SNPs that are missing >5% with `--geno`, and then we'll do the same for individuals with `--mind`. We'll constantly save our output to new files by the way, so you'll know which step was performed last for each set of files. 

```{r filter_miss, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_dryad \
  --geno 0.05 \
  --make-bed --out cr237_filter1
plink --dog --bfile cr237_geno \
  --mind 0.05 \
  --make-bed --out cr237_filter1
```

You can see how many variants were dropped in plink's output to the terminal (as well as other useful data for tracking our dataset):   
- 7468 variants removed due to missing genotype data (--geno).
- 0 dogs removed due to missing genotype data (--mind).

You'll see that we get warnings about '25567 het. haploid genotypes', 'Nonmissing nonmale Y'-  don't worry about that yet.



### b. Sex discrepency check

We can use the genotype data to check whether the recorded sex for each individual makes sense (mistakes in the lab can happen). This is based on looking at SNPs on the X chromosome, but the X has a large pseudo-autosomal region that we need to ignore first (6630000 - 126883977).
In the dog dataset we see some unusual results, so this entire step is shown as a demo and we won't delete and individuals or impute their sexes.

To ignore SNPs in the canine pseudo-autosomal region of X-chromosome

```{r splitX, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_filter1 \
  --split-x 6630000 126883977 \
  --make-bed --out cr237_splitx
```

568 chromosome codes changed here, which is consistent with the  565 SNPs in the PAR reported here: https://www.genetics.org/content/184/2/595#T3.
The number of heterozygous haploid genotypes decreases to 5284 as a result of updating the way plink handles the variants.

We will use F-values produced by `--check-sex` (metric of X chromosome homozygosity/ inbreeding) to inform us if any individuals are sexed incorrectly or ambiguously. We add arguments for the threshold for females and males to be flagged. These are set very loosely here; the default thresholds will OK males if F > 0.8 and females if F < 0.2. In visualizing F-value distributions, males should appear near 1, females more loosely distributed around 0. We make a list of flagged individuals with `grep` and `awk`.

```{r sex_check, engine = 'bash', eval = F, include = T}
plink --bfile cr237_splitx --dog \
  --check-sex 0.799 0.80 
  
grep "PROBLEM" plink.sexcheck | awk '{print$1,$2}' \
  > sex_discrepancy.txt

wc -l sex_discrepancy.txt 
cat sex_discrepancy.txt
```

The output tells us that 6 problems are detected with these thresholds. We plot the sex-check results in the R section below, where we see that there are some outliers in both sexes. We could make a list to remove those individuals, but Baker et al. did not filter any individuals here, so we'll do the same. 

One other option is to impute the sex of ambiguous individuals (`--impute-sex)`, as shown here. As before, the thresholds are given for labeling any problem cases. We do `--check-sex` again to get the .sexcheck results.

```{r impute_sex, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_splitx \
  --impute-sex 0.799 0.80   \
  --make-bed --out cr237_splitx_impute
  
plink --dog --bfile cr237_splitx_impute  \
  --check-sex 0.799 0.80 
  
grep "PROBLEM" plink.sexcheck # no more problems
```

Or we could remove the 6 individuals with the 'PROBLEM' status using the list we made.

```{r remove_discrep, engine = 'bash', eval = F, include = T}
plink --bfile cr237_splitx  --dog  \
  --remove sex_discrepancy.txt \
  --make-bed --out cr237_splitx_drop
```

We'll keep all subjects and the 6 imputed labels  (`cr237_splitx_impute`) for the next step. Perhaps the F-values for dogs can be bit more variable that the expectations outlined above which are suited to human genotype data.



### c. Autosomal SNPs only

This is pretty straightforward. Generally, we only want to analyze the autosomal SNPs, not the X, Y, or mitochondrial chromosomes. 

Here we use `awk` to output a list of SNPs that are on chromosomes 1-38 (dog autosomes) from the .bim file. The we use `--extract` to retrain only these SNPs in the output `cr237_filter2`.

```{r autosomes, engine = 'bash', eval = F, include = T}
awk '{ if ($1 >= 1 && $1 <= 38) print $2 }' cr237_splitx_impute.bim \
  > snp_1_38.txt

plink --dog --bfile cr237_splitx_impute \
  --extract snp_1_38.txt \
  --make-bed --out cr237_filter2
```

Great, we removed ~ 6k SNPs and now only have autosomal SNPs for our GWAS.



### d. MAF filtering  

Generally, we want a threshold of MAF > 0.05, or possibly lower but only if our sample is very large. We can use `--freq` to get the MAF statistics for all SNPs, and `--maf` to filter the dataset to a specified threshold (5%). 

```{r maf, engine = 'bash', eval = F, include = T}
plink --bfile cr237_filter1 --dog \
  --freq --out MAF_check

# Remove SNPs with MAF frequency <5%
plink --bfile cr237_filter2 --dog \
  --maf 0.05 \
  --make-bed --out cr237_filter3
```

We removed 45k SNPs with small minor allele frequency. The results from --freq (MAF_check.frq) can be plotted to show the minor allele frequency distribution (see the R section).



### e. Hardy-Weinberg equilibrium

Do we have far more heterozygous calls than expected? That could mean there is systemic calling error (bad). If we have fewer heterozygotes than expected, it could be due to population stratification (also bad if not controlled). HW-equilibrium means no migration, natural selection, and/or non-random mating violations. GWAS methods assume HW-equilibrium for allele frequencies, so it is important to remove any SNPs that are far from equilibrium.

First we use `--hardy` to output the HWE p-values to the file plink.hwe for plotting (see R section). Afterwards, we use `--hwe` to remove SNPs that deviate from HWE with p-value smaller than 1e-7 (as Baker *et al.* did). By default, `--hwe` only removes these based on the control group, so we repeat the call with the `--hwe-all` flag to remove them for the cases too. 

```{r HWE, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_filter3 --hardy 

plink --dog --bfile cr237_filter3 \
  --hwe 1e-7 \
  --make-bed --out cr237_filter4

plink --dog --bfile cr237_filter4 \
  --hwe 1e-7 --hwe-all \
  --make-bed --out cr237_filter4
```

These two steps removed 43 and 93 variants.

### f. SNP pruning

We will now 'prune' the set of SNPs to get a subset consisting of loci that are in approximate linkage equilibrium. This removes a bunch of SNPs that are highly correlated and thus provide little extra information for association testing.

Do this with `--indep-pairwise`, where three parameters `50 5 0.2` are given to scan a window size 50 kbp, to shift the window 5 SNPs at each step, and to use 0.2 as the pairwise $r^2$ threshold. 

```{r prune_snps, engine = 'bash', eval = F, include = T}
plink --bfile cr237_filter4 --dog \
  --indep-pairwise 50kb 5 0.2 --out indepSNP
```

The output file indepSNP.prune.in has the subset of independent SNPs to retain and .out has 41k excluded SNPs. We'll use the indepSNP.prune.in file to only consider the pruned set of SNPs in future steps (don't delete this).

### g. Heterozygosity

Now with the pruned set of SNPs we want to remove any individuals with heterozygosity more than 3 sd from the mean. We use `--het` to check compute this statistic.

```{r het_check, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_filter4 \
  --extract indepSNP.prune.in --het \
  --out prune_check
```



We need to do some computation here to find the individuals that are > 3 sd away from the mean, which is more easily done in R. Open nano and make an Rscript called 'identify_het_outliers.R' with the following code.
```{r write_het_script, engine = 'bash', eval = F, include = T}
nano identify_het_outliers.R
```

Put the following R code into the script. It filters the results to find any individuals too far from the mean.

```{r het_script_code}
df <- read.table("./prune_check.het", header = T)
df$het <-  (df$"N.NM." - df$"O.HOM.") / df$"N.NM."
mean_het <- mean(df$het)
sd_het <- sd(df$het)

fails <- subset(
  df, (df$het < mean_het - 3*sd_het) | (df$het < mean_het + 3*sd_het)
)
fails$dst <- (fails$het - mean_het) / sd_het
write.table(fails, "hetero_outliers.txt", row.names = FALSE)

```

Save the script and exit nano. Set permissions and run the Rscript with --no-save to avoid creating any workspace files.

```{r R_het_filter, engine = 'bash', eval = F, include = T}
chmod u+x identify_het_outliers.R 
Rscript --no-save identify_het_outliers.R 
```

This will generate the file hetero_outliers.txt.
We have to reformat the file with `sed` to remove the quotation marks that R adds to the header and `awk` to take only the FID and IID columns. The list then gets passed to plink `--remove` to take those individuals out.

```{r remove_het_outliers, engine = 'bash', eval = F, include = T}
sed 's/"// g' hetero_outliers.txt | awk '{print$1, $2}' \
  > hetero_outliers.txt

plink --dog --bfile cr237_filter4 \
  --remove hetero_outliers.txt \
  --make-bed --out cr237_filter5
```

Note that the list made by the Rscript is empty - we didn't have any individuals that needed to be removed but if we did, they would now be excluded.


### h. Cryptic relatedness / Identity by descent

Since an independant sample is assumed for association testing, we want to exclude any related individuals from analysis. 
Here we examine data for individuals with high relatedness and compute the pi_hat statistic for identity by descent. Normally, we remove any individuals having 'cryptic relatedness', where `pi_hat` is greater than some threshold. In their article, Baker *et al.* don't mention doing this at all, so we may get different results by applying this filter.

The `--genome` function is used to compute relation between pairs of individuals (identity by descent). The `--min` flag sets minimum PI_HAT for the output (otherwise we get all values). A PI_HAT of 1 implies identical individuals (duplicates in data), 0.5 is parent-child or siblings, 0.25 is aunt/uncle-niece/nephew. We want to scrutinize anything over 0.2 as these pairs could be related.

```{r genome_pihat, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_filter5 \
  --extract indepSNP.prune.in \
  --genome  --min 0.2 \
  --out pihat_min0.2
  
wc -l pihat_min0.2.genome
```

82 pairs of dogs have pi_hat > 0.2. If we were being rigorous, we'd probably want to remove one individual from each pair (see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/ for a more in-depth explanation). Otherwise, we might account for known relationships in the association model (as per Baker *et al.*) but that is beyond our scope. Here we remove one individual from a pair that are likely parent-child or siblings. 

We can filter the .genome results on Z1 values to find any pairs that likely represent siblings (Z1 > 0.9) as follows:

```{r siblings, engine = 'bash', eval = F, include = T}
awk '{ if ($8 >0.9) print $0 }' pihat_min0.2.genome > \
  siblings.genome
cat siblings.genome
```

This pair of dogs could be siblings or parent-child based on their Z1 value. We should remove one of these individuals from our analysis, preferably the one with less missing data. 

We'll reshape a list of ids from the siblings.genome file and get the missing data for these individuals

```{r sib_missing, engine = 'bash', eval = F, include = T}
grep -v "FID" siblings.genome > siblings.genome2
awk '{print$1, $2}' siblings.genome2 | uniq > siblings.txt
awk '{print$3, $4}' siblings.genome2 | uniq >> siblings.txt

plink  --dog --bfile cr237_filter5 \
  --keep siblings.txt \
  --missing \
  --out siblings 

head siblings.imiss
```

We could write code to select the individual with less missing data programmatically, but since there is only one pair, we can easily tell that individual 1251 has 1% more missing data than 1250.
Write "1251 1251" in a text file called 'drop_this.txt' and pass it to --remove to drop this individual from our data.

```{r, engine = 'bash', eval = F, include = T}
plink --dog --bfile cr237_filter5 \
  --remove drop_this.txt \
  --make-bed --out cr237_filter6
```



**We are finally done filtering!**  

This is a good time to download the plink.* files to your local machine for plotting.
We generated a lot of intermediate files, so I suggest removing these to keep the workspace tidy. You can safely remove these as follows

```{r, engine = 'bash', eval = F, include = T}
for i in 1 2 3 4 5 ; do rm *_filter$i.*; done
rm sex_discrepancy.txt snp_1_38.txt \
  *splitx* het_fail_ind.txt hetero_outliers.txt \
  missing2* siblings* drop_this.txt
```


---

## Multidimensional scaling

If we have resources to place our study subjects within their  'ethnic' groups (eg. 1000 genomes SNPs data for human ethnicity), then we can 'anchor' our samples with this data to find out where the samples lie in relation to these groups. 

This uses MDS to show distance between individuals, such that we can identify any individuals that are outliers from their groups. Because we want to do our study on a homogeneous population, we would remove any samples that deviate from the group based on a selected threshold.

Since we are dealing with Labrador retrievers, we'll just perform MDS to check for any outliers. We also take the MDS components as covariates to control for any stratification in our sample

For more details about anchoring your data, [click here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6001694/), to see an example with human SNP data anchored by the 
1000 genomes data.
There is more information about [plink clustering options here](https://www.cog-genomics.org/plink/1.9/strat). 


We'll use the set of files named 'cr237_filter6.*' for MDS in the following code.



---

## Association analyis

### Chi-sq test

### Logistic regression model

### Adjusted statistics

### Permutation

---

## Polygenic Risk Scores


----


## Visualizing the results


```{r libraries, include=F, eval=T}
library(tidyverse)  # 
library(patchwork)
theme_set(theme_bw()) # this sets the base theme for all your plots
```



### QC: Missing data

```{r missing_hist, fig.width=6, fig.height=3}
# plots histogram of missing data for individuals from .imiss file
gwa_missing_indiv <- 
  function(plink_imiss){
    plink_imiss %>% 
      janitor::clean_names() %>% 
      transmute(f_miss = as.numeric(f_miss)) %>% 
      ggplot(aes(f_miss)) +
      geom_histogram() +
      labs(subtitle = "Missing data per individual", 
           x = 'proportion missing',
           y = 'individuals')
  }

# plots histogram of missing data for SNPs from .lmiss file
gwa_missing_snp <- 
  function(plink_lmiss){
    plink_lmiss %>% 
      janitor::clean_names() %>% 
      transmute(f_miss = as.numeric(f_miss)) %>% 
      ggplot(aes(f_miss)) +
      geom_histogram() +
      labs(subtitle = "Missing data per SNP", 
           x = 'proportion missing',
           y = 'SNPs')
  }

# read data and make plots
indiv <- read_delim("./plink.imiss", delim = ' ') %>% 
  gwa_missing_indiv()
snps <- read_delim("./plink.lmiss", delim = ' ') %>% 
  gwa_missing_snp()

# with patchwork:: + operator to combine plots
indiv + snps
```


### QC: The Sex-check

```{r sex_check_hist, fig.height=3, fig.width=4.5}

# plots plink --sex-check results
gwa_sexcheck <- function(plink_sexcheck, plt_title){
  plink_sexcheck %>% 
  janitor::clean_names() %>% 
  transmute(f_value = as.numeric(f),
            pedsex = as.numeric(pedsex),
            pedsex = case_when(
              pedsex == '1' ~ "male",
              pedsex == '2'~ "female",
              TRUE ~ "labeled as ambiguous"),
            ) %>% 
  ggplot(aes(x = f_value)) +
  geom_histogram() +
  facet_wrap(~pedsex, ncol = 2) +
  theme(legend.position = 'NA')
}

# read and plot data; you can add more layers to ggplots to customize
# using + labs(), + theme(), etc.
read_delim("./plink.sexcheck", delim = ' ') %>% 
  gwa_sexcheck() +
  labs(
    x = 'X-chromsome inbreeding coefficient',
    y = 'dogs',
    subtitle =  "--sex-check: input data vs SNP-imputed sexes. 
    Outliers could be labelled incorrectly in the input data"
    ) 
```
### Minor allele frequency from plink --freq

```{r maf_hist}
read_delim("./MAF_check.frq", delim = ' ') %>% 
  janitor::clean_names() %>% 
  ggplot(aes(x = as.numeric(maf))) +
  geom_histogram() +
  labs(x = 'MAF', y = 'SNPs', 
       subtitle = "Minor allele frequency distribution")
```

### HWE

```{r}
read_delim("./plink.hwe", ' ') %>% 
  janitor::clean_names() %>% 
  transmute(p = as.numeric(p)) %>% 
  mutate(fill = if_else(p < 1e-07, '< 1e-7', 'keep')) %>% 
  ggplot(aes(x = p, fill = fill)) +
  geom_histogram() +
  labs(x = 'p-value', y = 'SNPs', fill = '',
       subtitle = "Hardy-Weinberg equilibrium test")
```

